# Microbiome Analysis in R

Repository for the course "Microbiome Analysis in R"

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.13650233.svg)](https://doi.org/10.5281/zenodo.13650233)


## Summary

This course aims to provide a comprehensive understanding of microbiome analysis using R, specifically through the *microeco* package. Participants will learn how to load, manipulate, and normalize microbiome data, as well as calculate relative abundances, perform alpha and beta diversity analysis, and generate graphical representations. Using Google Colaboratory as the platform, the course is accessible to all levels, from beginners to professionals, with no prior bioinformatics experience required. Additionally, practical learning is promoted with statistical tools that facilitate the interpretation and analysis of microbiological data.

## Learning Objectives

By the end of this course, participants will be able to:

- Load and prepare microbiome data generated by Qiime2 in R using the *microeco* package.
- Manipulate and normalize microbiological data, including calculating relative abundances.
- Perform alpha and beta diversity analysis to assess the variability of bacterial composition in the samples.
- Apply statistical tools to interpret the results of diversity analysis and perform hypothesis testing.
- Generate graphical representations such as bar plots, box plots, and heatmaps to visualize microbiological composition and diversity.
- Interpret results in the context of microbiome analysis using visualizations and descriptive statistics.

## Target Audience

The course is aimed at students, researchers, or clinical/sanitary professionals with prior knowledge of Qiime2 and basic R knowledge.

## General Requirements

The course will use Google Colaboratory, which is free, but requires a Google account to use. Google accounts are free.

## Acknowledgments

We thank Daniel López Juárez, who developed this course, for his valuable contribution and effort in creating this educational material.

## Program

### Module 1: [Introduction to Google Colaboratory]()

To get started, go to https://colab.research.google.com/.

On the Colab homepage, select "File" and then "Open notebook". 

There is a tab for "GitHub"; select that tab and paste the following URL in the search bar under "Enter a GitHub URL or search by organization or user":

`https://github.com/cabana-online/Tutorial_Introduccion`

After a brief search, you will see the notebook:

`01.Intro_a_colab.ipynb`

Select it, and the notebook will open.

To run the cells, you will need to log in with your Google account (it’s free to create one). The use of Colab is free. 

**Note:** There are limitations in the free version, but they will not be reached in this course.

### Module 2: [Metagenome Analysis in R]()

To get started, go to https://colab.research.google.com/.

On the Colab homepage, select "File" and then "Open notebook". 

There is a tab for "GitHub"; select that tab and paste the following URL in the search bar under "Enter a GitHub URL or search by organization or user":

`https://github.com/cabana-online/Tutorial_Introduccion`

After a brief search, you will see the notebook:

`01.Intro_a_colab.ipynb`

Select it, and the notebook will open.

To run the cells, you will need to log in with your Google account (it’s free to create one). The use of Colab is free. 

**Note:** There are limitations in the free version, but they will not be reached in this course.
